<a href="https://colab.research.google.com/github/SophistryDude/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/Nick_M's_Copy_of_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [1]:
# YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-08-11 19:48:23--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.07s   

2020-08-11 19:48:24 (284 KB/s) - ‘student.zip’ saved [20478/20478]



In [2]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats

In [5]:
df = pd.read_csv('student-mat.csv', sep=";")
df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [101]:
# For my first test I'm going to look at abscences and failures as having a strong association. to do so we create a crosstab to analyze the data
dependent1 = pd.crosstab(df['failures'], df['absences'])

dependent1

absences  0   1   2   3   4   5   6   7   ...  26  28  30  38  40  54  56  75
failures                                  ...                                
0         89   3  58   6  47   4  25   6  ...   1   0   1   0   0   1   1   1
1         14   0   4   1   4   0   3   1  ...   0   1   0   1   1   0   0   0
2          5   0   1   0   2   0   2   0  ...   0   0   0   0   0   0   0   0
3          7   0   2   1   0   1   1   0  ...   0   0   0   0   0   0   0   0

[4 rows x 34 columns]

# Are abscences and failure rates associated #
## Null Hypothesis ##
Abscences and Failure rates are independent
## Alternate Hypothesis ##
Abscences and Failure rates are dependent

In [102]:
chi2, p_value, dof, expected = stats.chi2_contingency(dependent1)
print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies:", expected)

chi2: 122.18899006084003
p value: 0.05694057065252934
dof: 99
expected_frequencies: [[9.08354430e+01 2.36962025e+00 5.13417722e+01 6.31898734e+00
  4.18632911e+01 3.94936709e+00 2.44860759e+01 5.52911392e+00
  1.73772152e+01 2.36962025e+00 1.34278481e+01 2.36962025e+00
  9.47848101e+00 2.36962025e+00 9.47848101e+00 2.36962025e+00
  5.52911392e+00 7.89873418e-01 3.94936709e+00 7.89873418e-01
  3.15949367e+00 7.89873418e-01 2.36962025e+00 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01]
 [1.45569620e+01 3.79746835e-01 8.22784810e+00 1.01265823e+00
  6.70886076e+00 6.32911392e-01 3.92405063e+00 8.86075949e-01
  2.78481013e+00 3.79746835e-01 2.15189873e+00 3.79746835e-01
  1.51898734e+00 3.79746835e-01 1.51898734e+00 3.79746835e-01
  8.86075949e-01 1.26582278e-01 6.32911392e-01 1.26582278e-01
  5.06329114e-01 1.26582278e-01 3.79746835e-01 1.26582278e-01
  1.26582278e-0

# Conclusion #
Based on the chi2 statistics of 122.189 and a p value of .795 we reject the null hypothesis that abscences and failures are independent. 


In [99]:
# There isn't a strict dichotomy between these two categories, but it's close. This should give us an association
dependent2 = pd.crosstab(df['schoolsup'], df['famsup'])
dependent2

famsup      no  yes
schoolsup          
no         140  204
yes         13   38

# Are family supplies and school supplies associated #
## Null Hypothesis ##
family supplies and school supplies are independent
## Alternate Hypothesis ##
family supplies and school supplies are dependent



In [100]:
chi2, p_value, dof, expected = stats.chi2_contingency(dependent2)
print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies:", expected)

chi2: 3.711345590895456
p value: 0.05404380626633763
dof: 1
expected_frequencies: [[133.24556962 210.75443038]
 [ 19.75443038  31.24556962]]


# Conclusion #
Based on the chi2 statistics of 4.328 and a p value of .3634 we reject the null hypothesis that family supplies and school supplies are independent. 



In [16]:
df = pd.read_csv('student-mat.csv', sep=";")
df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

# Is address and age associated #
## Null Hypothesis ##
address and age are independent
## Alternate Hypothesis ##
address and age are dependent




In [107]:
# Let's create a new table that focuses on our columns
dependent3 = pd.crosstab(df['age'], df['address'])
dependent3

address   R   U
age            
15       15  67
16       11  93
17       27  71
18       27  55
19        7  17
20        0   3
21        1   0
22        0   1

In [98]:
# There is a problem with analyzing the data like this, but let's run the chi^2 test anyways
chi2, p_value, dof, expected = stats.chi2_contingency(dependent3)
print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies:", expected)

chi2: 21.212531837116867
p value: 0.0034680030378527907
dof: 7
expected_frequencies: [[18.26835443 63.73164557]
 [23.16962025 80.83037975]
 [21.83291139 76.16708861]
 [18.26835443 63.73164557]
 [ 5.34683544 18.65316456]
 [ 0.66835443  2.33164557]
 [ 0.22278481  0.77721519]
 [ 0.22278481  0.77721519]]


# Conclusion #
Based on the chi2 statistics of 21.213 and a p value of 0.0035 we fail toreject the null hypothesis that address and age are independent.

# Is G1 and G2 associated #
## Null Hypothesis ##
G1 and G2 are independent
## Alternate Hypothesis ##
G1 and G2 are dependent

In [106]:
dependent4 = pd.crosstab(df['G1'], df['G2'])
chi2, p_value, dof, expected = stats.chi2_contingency(dependent4)
print("chi2:", chi2)
print("p value:", p_value)

chi2: 1140.1317989446097
p value: 1.1382453405189429e-111


# Conclusion #
Based on the chi2 statistics of 1140.1318 and a p value of 1.138 x 10^111 we fail to reject the null hypothesis that address and age are independent.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [112]:
chi2, p_value, dof, expected = stats.chi2_contingency(dependent1)
print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies:", expected)

chi2: 122.18899006084003
p value: 0.05694057065252934
dof: 99
expected_frequencies: [[9.08354430e+01 2.36962025e+00 5.13417722e+01 6.31898734e+00
  4.18632911e+01 3.94936709e+00 2.44860759e+01 5.52911392e+00
  1.73772152e+01 2.36962025e+00 1.34278481e+01 2.36962025e+00
  9.47848101e+00 2.36962025e+00 9.47848101e+00 2.36962025e+00
  5.52911392e+00 7.89873418e-01 3.94936709e+00 7.89873418e-01
  3.15949367e+00 7.89873418e-01 2.36962025e+00 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01]
 [1.45569620e+01 3.79746835e-01 8.22784810e+00 1.01265823e+00
  6.70886076e+00 6.32911392e-01 3.92405063e+00 8.86075949e-01
  2.78481013e+00 3.79746835e-01 2.15189873e+00 3.79746835e-01
  1.51898734e+00 3.79746835e-01 1.51898734e+00 3.79746835e-01
  8.86075949e-01 1.26582278e-01 6.32911392e-01 1.26582278e-01
  5.06329114e-01 1.26582278e-01 3.79746835e-01 1.26582278e-01
  1.26582278e-0

In [114]:
dependent1_with_margin = pd.crosstab(df['failures'], df['absences'], margins=True)
dependent1_with_margin

absences    0  1   2  3   4  5   6  7   8  ...  26  28  30  38  40  54  56  75  All
failures                                   ...                                     
0          89  3  58  6  47  4  25  6  17  ...   1   0   1   0   0   1   1   1  312
1          14  0   4  1   4  0   3  1   2  ...   0   1   0   1   1   0   0   0   50
2           5  0   1  0   2  0   2  0   2  ...   0   0   0   0   0   0   0   0   17
3           7  0   2  1   0  1   1  0   1  ...   0   0   0   0   0   0   0   0   16
All       115  3  65  8  53  5  31  7  22  ...   1   1   1   1   1   1   1   1  395

[5 rows x 35 columns]

In [116]:
312*115/395

90.83544303797468

In [117]:
row_totals = df['failures'].value_counts().values 

row_totals

array([312,  50,  17,  16])

In [127]:
column_totals = df['absences'].value_counts().sort_index().values 
column_totals

array([115,   3,  65,   8,  53,   5,  31,   7,  22,   3,  17,   3,  12,
         3,  12,   3,   7,   1,   5,   1,   4,   1,   3,   1,   1,   1,
         1,   1,   1,   1,   1,   1,   1,   1])

In [128]:
sample_size = df.shape[0]
expected = np.array([[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
                     [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
                     [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
                     [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]])

for row_index, row in enumerate(row_totals):
  for column_index, col in enumerate(column_totals):
    expected_value = (row*col)/sample_size
    expected[row_index][column_index] = expected_value
expected

array([[9.08354430e+01, 2.36962025e+00, 5.13417722e+01, 6.31898734e+00,
        4.18632911e+01, 3.94936709e+00, 2.44860759e+01, 5.52911392e+00,
        1.73772152e+01, 2.36962025e+00, 1.34278481e+01, 2.36962025e+00,
        9.47848101e+00, 2.36962025e+00, 9.47848101e+00, 2.36962025e+00,
        5.52911392e+00, 7.89873418e-01, 3.94936709e+00, 7.89873418e-01,
        3.15949367e+00, 7.89873418e-01, 2.36962025e+00, 7.89873418e-01,
        7.89873418e-01, 7.89873418e-01, 7.89873418e-01, 7.89873418e-01,
        7.89873418e-01, 7.89873418e-01, 7.89873418e-01, 7.89873418e-01,
        7.89873418e-01, 7.89873418e-01],
       [1.45569620e+01, 3.79746835e-01, 8.22784810e+00, 1.01265823e+00,
        6.70886076e+00, 6.32911392e-01, 3.92405063e+00, 8.86075949e-01,
        2.78481013e+00, 3.79746835e-01, 2.15189873e+00, 3.79746835e-01,
        1.51898734e+00, 3.79746835e-01, 1.51898734e+00, 3.79746835e-01,
        8.86075949e-01, 1.26582278e-01, 6.32911392e-01, 1.26582278e-01,
        5.06329114e-01,

In [130]:
observed = dependent1.values

In [132]:
chi2 = ((observed - expected)**2 / expected).sum()
chi2

122.18899006084003

In [133]:
DoF = (4-1)*(34-1)
DoF

99

using the applet: https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html
we get a p val of 0.05694

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [110]:
# YOUR WORK HERE
crosstab = {}
for column_headers in df:
  if column_headers != '':
    crosstab[column_headers] = pd.crosstab(df[column_headers], df[column_headers])

  

In [84]:
Chi2 = {}
for column_headers in crosstab:
  if column_headers != '':
    Chi2[column_headers] = chi2, p_value, dof, expected = stats.chi2_contingency(dependent3)
print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies:", expected)

chi2: 3.342736233142963
p value: 0.9996529931940035
dof: 16
expected_frequencies: [[5.83341772e+01 7.39848101e+01 6.97164557e+01 5.83341772e+01
  1.70734177e+01 2.13417722e+00 7.11392405e-01 7.11392405e-01
  2.81000000e+02]
 [2.36658228e+01 3.00151899e+01 2.82835443e+01 2.36658228e+01
  6.92658228e+00 8.65822785e-01 2.88607595e-01 2.88607595e-01
  1.14000000e+02]
 [8.20000000e+01 1.04000000e+02 9.80000000e+01 8.20000000e+01
  2.40000000e+01 3.00000000e+00 1.00000000e+00 1.00000000e+00
  3.95000000e+02]]




### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"
A null hypothesis states that there is no relationship between the two variables, they are independent. 
- What is a p-value and what does it represent?
the p-value is the statistical likelihood of seeing a result as extreme. It represents how close the mean of your chi^2 test is compared to a normal distribution. 
- What does it mean for something to be "statistically significant?"
Something is statistically significant when the p value is small enough to note a dependent relationship.
- What is a test statistic and how does it relate to a p-value?
A test statistic is a random sampling that is used to determine whether you agree with the null statement. If the test statistic corroborates the larger data sample the p-value should be close to 0. Using a confidence rating of 95%, the p-value would be below 0.05
- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?
A degree of freedom is the number of axioms -1. So a 1-sample it would be the number of things tested -1, for a 2-sample it would be the summation of the 2 columns tested -2. The chi^2 test is (a-1)*(b-1)

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)